### PR TITLE
Stop query if user Id & api not set. Slowed down no chat windows notifications

### PR DESCRIPTION
**Pull request by cTheDragons**
Fix for Issue #38

+ Do not start querying if UserId and API token are not set (ie are each 36 length long)
+ Stop checking for notifications if chat is not active, user ids not set and if chat windows are open
+ Increase notifications requery to 40 seconds * by factor greater than hour.

### CODE REVIEW

**Comment by cTheDragons** [comment]
Have asked @Alys in the guild on the best method of testing this.

**Comment by cTheDragons** [comment]
Thanks for your patience with the multiple commits. I forgot to update the firefox zip and only realised when getting ready for bed. Should be done now till the testers have completed testing.

File: README.md [hunk 1]
```diff
@@ -24,6 +24,11 @@ To allow for ease of deployment the code is kept in separate folders, chrome, an
 
 # Version history
 
+### 2.1.1
```
Comment (paglias) [inline README.md]: could this actually be 2.1.2? because I had to publish 2.1.1 to disable the ajax calls

**Comment by cTheDragons** [comment]
Done. Updated the mainfest, zip and Readme

**Comment by cTheDragons** [comment]
Just a little update, I have been running some testing on the refresh rates on chat and have found if you extend the chat greater than 60 minutes it is still querying every 5 seconds if you just leave your computer open and the chat focused. I think I will need to adjust this. 

@Alys et all, could you confirm are you happy with the refresh rates for timeouts less than 60 minutes?

+ 5 seconds if the chat windows is focus (RefreshRateFast)
+ 45 seconds if the chat window is minmised (RefreshRateMedium) *I think*
+ 60 seconds if the chat window not focused (RefreshRateSlow)
+ 40 seconds if there is no chat windows.

I am planning to change the code if the timeout is greater than 60 minutes that it will proportionally increase the timeouts. Ie if you change the timeout to 120 minutes the RefreshRateFast will be every 10 seconds.

**Comment by paglias** [comment]
5 seconds is too frequent unfortunately, refresh rates should be of at
least 60s

Il Mar 12 Mar 2019, 00:25 cTheDragons <notifications@github.com> ha scritto:

**Comment by cTheDragons** [comment]
Really?

It has been like this [for a long time](https://github.com/HabitRPG/habitica-chat-extension/blob/a71dddd0246dd89b0af016dca7c9216d9d12deb7/mainChat/chat_inPage.js#L386). I think since the [beginning](https://github.com/HabitRPG/habitica-chat-extension/blob/51b6389c1a7c08fff617f6145c449a2792834b4a/chat_inPage.js#L319)... 

Should there be some API limiting if 5s calls are too much.

**Comment by paglias** [comment]
Eh, yeah there should be some rate limiting but right now we don't have
anything in place.

So in the meantime 5s is way too often given the server problems we're
having. Habitica chat is not meant to be realtime, I understand 60s might
be a little too slow, so let's make it 30s minimum if you want

(I also noticed you poll the entire user every time, would it be possible
to only fetch the notifications?)

Il Mar 12 Mar 2019, 02:03 cTheDragons <notifications@github.com> ha scritto:

**Comment by cTheDragons** [comment]
Will do. How do you fetch notifications?

The [API Doc only has posts options](https://habitica.com/apidoc/#api-Notification-ReadNotifications)

**Comment by paglias** [comment]
Sorry, I thought there was a route to only read notifications but looks
like I got confused!

Il Mar 12 Mar 2019, 09:32 cTheDragons <notifications@github.com> ha scritto:

**Comment by cTheDragons** [comment]
Is there are nice route other than user that will allow me to read notifications other than user?

**Comment by paglias** [comment]
No but you can use the userFields parameter to only get the needed fields
instead of the entire user

Il Mar 12 Mar 2019, 09:42 cTheDragons <notifications@github.com> ha scritto:

**Comment by cTheDragons** [comment]
All done. *(Including the readme file)*

+ Have set the refresh rate to 30s
+ Querying only achievements now instead of the entire user.
and other bits and bobs I found scrambling through the code. 

Will submit it back to the guild for testing.

**Comment by Alys** [comment]
@paglias there's a group of up to 15 socialites and mods who are available to test this. Is it okay if testing happens now or do you want to set up any monitoring in advance, or wait for anything else?

**Comment by paglias** [comment]
Testing now is fine

Il Mar 12 Mar 2019, 12:21 Alys <notifications@github.com> ha scritto:

**Comment by cTheDragons** [comment]
This can be released. 

Confirmed by the following users:
+ @citrusella on Firefox
+ @LtCabel on Chrome
+ @ravenlune  on Chrome

Let me know if you require anything else.

Thanks

**Comment by paglias** [comment]
thanks @cTheDragons ! There's just one thing we'd like you to add before this can be released: when making calls to the API, could you pass another header (alongside the x-api-user and x-api-key) to make it easier to identify requests coming from the chat extension? Something like `x-api-source` with a value of `chat-extension` should work

**Comment by paglias** [comment]
Actually regarding the header name, let's use `x-client`

**Comment by cTheDragons** [comment]
Changes completed. Just waiting on testing to occur with Socialites.

**Comment by cTheDragons** [comment]
Found some issues. These have been corrected. Retesting with 2.2.1.

**Comment by cTheDragons** [comment]
Version 2.2.1 ready to release.

This solves issue #38. 

Confirmation from 
+ @LtCabel  (Tier 7 Socialite, Artisan) Chrome Version 73.0.3683.103 (Official Build) (64-bit), Windows 10 Tier 7 Socialite, Artisan
+ @weeWitch (Tier 7 Linguist, Socialite, Artisan) Testing on Firefox Quantum 66.0.3 (64-bit) on a Windows 7 environment.
+ @aspiring_advocate (Tier 5 Artisan, Linguist, Socialite) Testing Chrome, Version 73.0.3683.103 on a Mac?
+ @aspiring_advocate (Tier 5 Artisan, Linguist, Socialite) Testing Firefox, Version Unknown on a Mac
+ @Ceran (Tier 4 Challenger, Socialite) Testing on Chrome, Version 73.0.3683.86, on a windows 7 laptop.

It should be noted that both with @aspiring_advocate on Firefox and @Ceran on Chrome both encountered issues 
+ Linking the API Key and Token cause the page to reload several times
+ Group List failed to load.

However in both these cases it is believe this was due to server issues at the time, as further tests this did not occurred.  (The system no longer re-queries if it cannot load the group list which explains this behaviour).

**Comment by paglias** [comment]
Thanks @cTheDragons ! Will briefly test locally and release

**Pull request merged**